src/robot/romi-robot.ts

```diff
@@ -566,43 +707,116 @@ export default class WPILibWSRomiRobot extends WPILibWSRobotBase {
+        this._dioDevicePortMapping = [];
+        this._analogInDevicePortMapping = [];
+        this._pwmDevicePortPortMapping = [];
+
+        // Clear out the External IO Pin config
         this._extPinConfiguration = [];
 
+        // Set up all onboard channels
+        ROMI_ONBOARD_DIO.forEach((val, idx) => {
+            this._dioDevicePortMapping.push({
+                device: "romi-onboard",
+                port: idx
+            });
+        });
+
+        ROMI_ONBOARD_PWM.forEach((val, idx) => {
+            this._pwmDevicePortPortMapping.push({
+                device: "romi-onboard",
+                port: idx
+            });
+        });
+
+        // Now configure the external IO
         this._ioConfiguration.forEach((pinConfig, ioIdx) => {
             switch (pinConfig.mode) {
                 case IOPinMode.ANALOG_IN:
                     this._extPinConfiguration.push(2);
-                    this._extAnalogInPins.push(ioIdx);
-                    // We also need to add these to the input set
-                    this._analogInputValues.set(this._extAnalogInPins.length - 1, 0.0);
+
+                    this._analogInDevicePortMapping.push({
+                        device: "romi-external",
+                        port: ioIdx // We use ioIdx here so that we know which offset to write to
+                    });
+
+                    this._analogInputValues.set(this._analogInDevicePortMapping.length - 1, 0.0);
                     break;
                 case IOPinMode.DIO:
                     // Default to OUTPUT for digital pins
                     this._extPinConfiguration.push(0);
-                    this._extDioPins.push(ioIdx);
+
+                    this._dioDevicePortMapping.push({
+                        device: "romi-external",
+                        port: ioIdx
+                    });
                     break;
                 case IOPinMode.PWM:
                     this._extPinConfiguration.push(3);
-                    this._extPwmPins.push(ioIdx);
+                    this._pwmDevicePortPortMapping.push({
+                        device: "romi-external",
+                        port: ioIdx
+                    });
                     break;
             }
         });
 
-        return this._writeIOConfiguration();
+        // Write the onboard and external IO configurations
+        // and then set up the custom devices
+        return this._writeRomiOnboardIOConfiguration()
+        .then(() => {
+            return this._writeRomiExtIOConfiguration();
+        })
+        .then(() => {
+            // Configure any custom devices we might have
+            this._customDevices.forEach(device => {
+                const ioInterfaces = device.ioInterfaces;
+
+                if (ioInterfaces.numDioPorts !== undefined) {
+                    for (let i = 0; i < ioInterfaces.numDioPorts; i++) {
+                        this._dioDevicePortMapping.push({
+                            device,
```

Comment:
Should this be device.identifier instead?

Reply:
Nope, the various `portMapping` arrays take in an object of type:

```typescript
interface DevicePortMapping {
  device: CustomDevice | "romi-onboard" | "romi-external",
  port: number
}
```

So handing in the actual device is correct